Docs/#1 english document (#20)
tier4/driving_log_replayer: docs/use_case/performance_diag.en.md

# Evaluate diagnostics

> Evaluate whether Autoware's diagnostics function as intended.

Reviewer: ```suggestion
Evaluate whether Autoware's diagnostics functionality behaves as expected.
```

> Currently, it supports evaluation of lidar visibility and blockage.

Reviewer: ```suggestion
The following subjects of evaluation are currently supported:
```

- visibility: function to determine if visibility is impaired by fog, rain, etc.
- blockage: function to determine if leaves or other objects are attached to LiDAR and obstructing the measurement.

## Evaluation Method

> Use performance_diag.launch.py to evaluate.
> When launch is launched, the following is executed and evaluated.
> 
> 1. launch the evaluation node (performance_diag_evaluator_node), logging_simulator.launch and ros2 bag play
> 2. autoware receives sensor data output from bag and outputs /diagnostics_agg
> 3. evaluation node subscribe to /diagnostics_agg, evaluates in callback and records the result in a file
> 4. when the playback of the bag is finished, launch is automatically terminated and the evaluation is finished

Reviewer: ```suggestion
The diagnostics evaluation is executed by launching the `performance_diag.launch.py` file.
Launching the file executes the following steps:

1. Execute launch of evaluation node (`performance_diag_evaluator_node`), `logging_simulator.launch` file and `ros2 bag play` command.
2.Autoware receives sensor data output from the input rosbag and outputs the`/diagnostics_agg` topic.
3. The evaluation node subscribes to `/diagnostics_agg` topic,  and evaluates data. The result is dumped into a file.
4. When the playback of the rosbag is finished, Autoware's launch is automatically terminated, and the evaluation is completed.
```

### visibility evaluation

> In the evaluation of visibility, it is confirmed that more than a certain number of ERRORs for visibility are output using data obtained in rainy weather or at facilities where rain is artificially generated.
> Also, using data obtained during sunny weather, it is confirmed that ERRORs are never generated.

Reviewer: ```suggestion
The evaluation process confirms that more than a specified number of ERRORs for limited visibility are generated for rosbag input data obtained under rainy conditions (naturally or artificially generated).
Also, using data obtained during sunny weather, it is confirmed that ERRORs are never generated.
```

### blockage evaluation

> For blockage evaluation, acquire data with LiDAR intentionally covered with a plastic bag, etc., and confirm that ERROR for blockage is output more than a certain number of times.
> Also, confirm that no ERROR is generated for LiDAR that is not covered with a plastic bag.

Reviewer: ```suggestion
For blockage evaluation, acquire data with LiDAR intentionally covered with a material that will not pass the laser beam (for example box). The evaluation confirms that ERROR for blockage is output more than a certain number of times for the considered situation.
The node will also confirm that no ERROR is generated for not covered LiDAR.
```

## Evaluation Result

> For each subscribe of LiDAR diagnostic results, the judgment result described below will be output.

Reviewer: ```suggestion
For each LiDAR diagnostic subscription, the evaluation judgment will be published on the topics described below:
```

> - visibility: /autoware/sensing/lidar/performance_monitoring/visibility/. \*.
> - blockage: /autoware/sensing/lidar/performance_monitoring/blockage/. \*

Reviewer: Should the `/. \*.` suffixes remain unchanged?

```suggestion
- visibility: `/autoware/sensing/lidar/performance_monitoring/visibility/*
- blockage: `/autoware/sensing/lidar/performance_monitoring/blockage/*`
```

> The case in which an ERROR is generated can be considered a success or a failure depending on what you want to evaluate, so you can change this by describing the type in the scenario.

Reviewer: ```suggestion
Each output of the evaluation can be considered a success or a  failure depending on what you want to evaluate. You can change this by describing the type in the scenario.
```

> - If the scenario type is TP, success will be achieved if a certain number of Diag ERRORs are generated.
> - If the scenario type is FP, it succeeds if Diag never becomes ERROR.
> - If the scenario type is null, the test is omitted.

Reviewer: ```suggestion
- If the scenario type is TP (true positive), success will be achieved if a certain number of `Diag` ERRORs is generated.
- If the scenario type is FP (false positive), it succeeds if the `Diag` state never becomes ERROR.
- If the scenario type is null, the test is omitted.
```

### TP Normal

IF the scenario type is TP and the level of visibility or blockage in the diagnostic information (/diagnostics_agg) is ERROR (=2)

### TP Error

> If the scenario type is TP and the level of visibility or blockage in the diagnostic information (/diagnostics_agg) is not ERROR (! =2)

Reviewer: ```suggestion
If the scenario type is TP and the level of visibility or blockage in the diagnostic information (`/diagnostics_agg`) is not ERROR.
```

### FP Normal

> If the scenario type is FP and the level of visibility or blockage in the diagnostic information (/diagnostics_agg) is not ERROR (!=2) =2)

Reviewer: ```suggestion
If the scenario type is FP and the level of visibility or blockage in the diagnostic information (`/diagnostics_agg`) is not ERROR.
```

### FP Error

If the scenario type is FP and the level of visibility or blockage in the diagnostic information (/diagnostics_agg) is ERROR (=2)

## Topic name and data type used by evaluation node

> - subscribe

Reviewer: ```suggestion
Subscribed topics:
```

| Topic name                                   | Data type                             |
| -------------------------------------------- | ------------------------------------- |
| /perception/obstacle_segmentation/pointcloud | sensor_msgs::msg::PointCloud2         |
| /diagnostics_agg                             | diagnostic_msgs::msg::DiagnosticArray |
| /tf                                          | tf2_msgs/msg/TFMessage                |

> - publish

Reviewer: ```suggestion
Published topics:
```

| /driving_log_replayer/visibility/value                   | example_interfaces::msg::Float64              |
| /driving_log_replayer/visibility/level                   | example_interfaces::msg::Byte                 |
| /driving_log_replayer/blockage/{lidar_name}/ground/ratio | example_interfaces::msg::Float64              |
| /driving_log_replayer/blockage/{lidar_name}/sky/ratio    | example_interfaces::msg::Float64              |
| /driving_log_replayer/blockage/{lidar_name}/level        | example_interfaces::msg::Byte                 |
| /initialpose                                             | geometry_msgs::msg::PoseWithCovarianceStamped |

{lidar_name} contains the name of the mounted lidar.

## Arguments passed to logging_simulator.launch

To lighten autoware processing, modules that are not relevant to evaluation are disabled by passing false as a launch argument.
The following is set.

- planning: false
- control: false
- localization: false / true (default value is false. Specify in scenario)

## About simulation

State the information required to run the simulation.

### Topic to be included in the input rosbag

> The CAN of the vehicle's ECU and the topic of the sensor being used are required.
> The following is an example, and may be changed if different sensors are used.
> 
> If multiple LiDARs are installed, include packets for all LiDARs installed.
> 
> - /sensing/gnss/ublox/fix_velocity
> - /sensing/gnss/ublox/nav_sat_fix
> - /sensing/gnss/ublox/navpvt
> - /sensing/imu/tamagawa/imu_raw
> - /sensing/lidar/\*/velodyne_packets
> - /gsm8/from_can_bus

Reviewer: ```suggestion
The vehicle's ECU CAN and sensors data topics are required for the evaluation to be run correctly.
The following example shows the topic list available in evaluation input rosbag when multiple LiDARs are used in a real-world vehicle configuration.

- /sensing/gnss/ublox/fix_velocity
- /sensing/gnss/ublox/nav_sat_fix
- /sensing/gnss/ublox/navpvt
- /sensing/imu/tamagawa/imu_raw
- /sensing/lidar/\*/velodyne_packets
- /gsm8/from_can_bus
```

        front_lower: # Set for each on-board LiDAR
          ScenarioType: TP # TP/FP/null
          BlockageType: both # sky/ground/both Where is the blockage occurring?
          PassFrameCount: 100 # If ScenarioType is TP, Blockage type matches, and ERROR is generated at least this many times, the Blockage test is considered successful.  for FP, this condition is ignored because the test must not generate any ERRORs.
        front_upper:
          ScenarioType: TP
          BlockageType: both
          PassFrameCount: 100
        left_lower:
          ScenarioType: FP
          BlockageType: sky
          PassFrameCount: 30
        left_upper:
          ScenarioType: FP
          BlockageType: both
          PassFrameCount: 40
        rear_lower:
          ScenarioType: FP
          BlockageType: ground
          PassFrameCount: 50
        rear_upper:
          ScenarioType: FP
          BlockageType: sky
          PassFrameCount: 60
        right_lower:
          ScenarioType: FP
          BlockageType: both
          PassFrameCount: 70
        right_upper:
          ScenarioType: FP
          BlockageType: ground
          PassFrameCount: 80
```

### Evaluation Result Format

> In performance_diag, visibility and blockage are evaluated.
> The Result is true if both visibility and blockage are passed, and false otherwise.
> 
> The format is shown below.
> However, common parts that have already been explained in the result file format are omitted.

Reviewer: ```suggestion
In `performance_diag` evaluation scenario visibility and blockage are evaluated.
The `Result` is `true` if both visibility and blockage evaluation steps have passed. Otherwise, the `Result` is `false`.

The result format is shown below.
*NOTE: common part of the result file format, which has already been explained, is omitted.*

```
